### PR TITLE
Add Interactive Dashboard

### LINKED ISSUE #91: Visualisation of input data
Create standard charts to allow users to interactively explore model input data.

### PR DESCRIPTION
**Pull request by trevorb1**
# The tl;dr
@tniet, @abhishek0208, @maartenbrinkerink; I created a dashboard to visualize OG models! It is close to being done and I am looking for feedback! Are (as least one of) you able to try running the dashboard to see if it works on your computer and let me know what you think? No big rush at all though! :) 

Steps to run: 
1. Checkout the `dashboard` branch
2. Check your dependencies! There are a few new ones / updates to the env file. Notable changes are `otoole>=1.0.0`, and `geopandas` and `dash` have been added. 
3. Install osemosys global as a package with the command `pip install -e .` from the root folder. (I will deploy `osemosys_global` to PyPI before merging `develop` into `main` though, so this is only a temporary step). 
4. Run the workflow as normal with `snakemake -c`. My example below is a 8 timeslice model of BBIN from 2020-2070; I have not tried the dashboard with a really big model. 
5. Run the command ~~`snakemake dashboard -c`~~ `bash dashboard.sh`. This should start the dashboard in a local host that will display in your web browser. (Updated command based on comment thread) 

# PR Description

In this PR I have created an interactive dashboard to visualize the input and result data from an OSeMOSYS Global model run. The dashboard (built with `plotly.dash`) consists of 5 tabs; tab 1 is a options tab, and the remaining tabs are different visualization options (described below). To reuse some existing plotting functions, a refactor of the visualization module was done to separate data and logic. 

## Options Tab
Global plotting options for the dashboard. Of note is the `Geographic Scope` radio buttons. If the option is set to `System`, the legend axis in graphs will be by technology. If the option is set to `Country` or `Region`, the legend axis will be country or region respectively. 

### Example of Options tab for a BBIN model
image

## Geographic Overview 
Allows the user to visualize what nodes and transmission lines are (and are not) included in the model. The user can hover over each node/line to get the OG name. This is in response to feedback that users did not know how to name new transmission lines in the `user_defined_capacity` option in the configuration file. Note, this pulls the lats/lons from the raw data sources; this means custom nodes will not be displayed in this tab (the other tabs will correctly display custom node data though). 

### Example of Nodes
image

### Example of Lines
image

## Input Data Tab
Plots input data. The plot will take on the values described in the options tab. The tech/fuel dropdown is dynamic, based on the users parameter option. The parameter options are hard coded following a format shown below, and passed into a generic plotting function. This allows us to easily add/remove parameters and variables. The plotting options include `Area`, `Line`, `Stacked Bar` and `Grouped Bar`.

### Example of system level residual capacity area plot
image

## Result Data Tab 
Repeat of input data tab, except with result variables. 

### Example of coal production by country by timeslice in 2040
image

## Transmission Tab
Similar to result data tab, however, shows result values for individual transmission lines. 

### Example of transmission capacity of the line `TRNINDNOINDWE`
image

## Other Notes
- Im just realizing there is no plot `All` option for results by country/region. For example you can only see emission results for a specific technology when the legend values are countries. I will fix this soon. 

- Below is how the parameter configurations used for plotting are stored. Some of this data can be pulled from other sources as well (such as the default value from the `otoole` config)

```yaml
"SpecifiedAnnualDemand":{
    "nicename":"Annual Demand",
    "groupby":"FUEL",
    "groupby_method":"sum",
    "xaxis":"YEAR", 
    "ylabel":"DEMAND (PJ)",
    "add_default":True,
    "default":0
},
```

- The options tab does not currently save your selection once you flip back to it. For example, if I change the plot theme and go to the results tab, the theme will be saved. If I go back to the options tab, the theme will revert to the default value and I have to change it again. I am working to solve this. 

- The transmission flow does not show directional flow; you are able to see that the transmission line is being used, but can not tell in which direction the flow of energy is. 

- I tried my best to load as little data in at one time, but for graphs like the production by timeslice the dashboard can take a second to respond. Im unsure how this will scale with very big models. 

- I have slightly changed the folder structure of the scripts. This includes referencing all files through the package path (ie. `from osemosys_global.dashboard.utils import plot_data`). This requires osemosys global to be installed as a package. As noted above, I will deploy osemosys global to PyPI before these changes get merged into main. 

# Issue Ticket Number

Closes #91 

# Documentation

Instructions on the dashboard will be added to the docs in an upcoming commit.

### CODE REVIEW

**Comment by trevorb1** [comment]
Something is weird with the updated conda env file with this change :( I am looking into fixing it but in the mean time you can conda/pip install each package individually and that still works quite fast

**Comment by trevorb1** [comment]
Need to remove the i18n dependency. If you get a i18n import error, try running `pip install python-i18n[YAML]`. Will update comment once this dependency is removed. 

- [ ] Remove i18n dependency

**Comment by trevorb1** [comment]
To kill the dashboard, you press ctrl+C. This writes out a snakemake error. Look into how to fix this 

- [ ] Method to kill the dashboard nicely

**Review by tniet — COMMENTED** [review]
Hey @trevorb1 - really cool dashboard.  I think you caught the errors I had when running things.  With those small fixes it works well and looks really nice!

**Comment by tniet** [comment]
> Something is weird with the updated conda env file with this change :( I am looking into fixing it but in the mean time you can conda/pip install each package individually and that still works quite fast

I think you fixed this already for me as I could update the environment after your fix...

**Comment by trevorb1** [comment]
To be done before merging:

- [ ] Add "All" option to results tab when filtering by country /region
- [ ] Fix options saving when flipping back to options tab
- [ ] Implement directional transmission flow 
- [ ] Add docs on how to run the dashboard
